src/kernel/thread.c

```diff
@@ -585,7 +585,15 @@ void setNextInterrupt(void)
         next_interrupt = MIN(refill_head(NODE_STATE(ksReleaseHead)->tcbSchedContext)->rTime, next_interrupt);
     }
 
-    setDeadline(next_interrupt - getTimerPrecision());
+    assert(next_interrupt >= NODE_STATE(ksCurTime));
+
+    /* Either of the rTime values above might sit just after ksCurTime (less
+       than getTimerPrecision() in the future), so we need to make sure that
+       subtracting getTimerPrecision() does not put us into the past. We are
+       fine with scheduling an interrupt at ksCurTime, even though that is
+       strictly speaking also in the past (at kernel entry), because we assume
+       it will be raised immediately after we leave the kernel. */
+    setDeadline(MAX(NODE_STATE(ksCurTime), next_interrupt - getTimerPrecision()));
```

Comment:
I don't see why this distinction is made. If we assume setDeadline can handle times in the past, why not let it just set a time before ksCurTime? That is, why not just remove all the asserts instead of updating them?

If it makes verification easier, we can move the `- getTimerPrecision()` part into the setDeadline() functions.

Reply:
We could do that, we'd just have to remove the assertions about `ksCurTime` from the insides of the various `setDeadline` implementations. I didn't want to add that much code disturbance, but it would simplify things a bit. I'll give it a shot.

Apart from `ackDeadlineIRQ` in `arm_generic.h`, this is the only code that calls `setDeadline`.

Reply:
I have addressed these now (still need to squash things into coherent commits). In `arm_global_timer.h` we still have the following:

```c
    /* enable cmp */
    globalTimer->control |= BIT(COMP_ENABLE);
    /* if this fails PRECISION is too low */
    assert(getCurrentTime() < deadline || globalTimer->isr == 1u);
```

but I think that holds, because it basically says just after enabling the interrupt that it is either in the future or it has been raised just now, which is what we want.

Reply:
Yes, you're right.

It seems it depends on the hardware revision whether ARM global timer can handle times in the past:

https://developer.arm.com/documentation/ddi0407/g/Global-timer--private-timers--and-watchdog-registers/About-the-Global-Timer

> In r2p0 the comparators for each processor with the global timer fire when the timer value is greater than or equal to. In previous revisions the comparators fired when the timer value was equal to.

Revision 2 came out in 2009, only one year after Cortex-A9 was released, so it seems unlikely to be used by actual hardware released years later. The imx6 and zynq7000 platforms use the ARM global timer, both were released in 2011.

Reply:
- [ ] I should probably add a comment to that effect to the `assert`